llvm/test/CodeGen/AMDGPU/GlobalISel/srem.i32.ll

```diff
@@ -312,13 +312,13 @@ define <2 x i32> @v_srem_v2i32_pow2k_denom(<2 x i32> %num) {
 ; GISEL-NEXT:    v_sub_i32_e32 v0, vcc, v0, v4
 ; GISEL-NEXT:    v_sub_i32_e32 v1, vcc, v1, v3
 ; GISEL-NEXT:    v_subrev_i32_e32 v3, vcc, s4, v0
-; GISEL-NEXT:    v_subrev_i32_e32 v4, vcc, s4, v1
+; GISEL-NEXT:    v_subrev_i32_e32 v4, vcc, 0x1000, v1
```

Comment:
Why is the constant only folded into the second v_subrev, not both?

Reply:
I think the root cause is that some shrinking only happens post-RA (e.g. shrinking V_SUBREV_CO_U32_e64 only if VCC was allocated for the carry-out operand) which is too late to do any more folding.

And the reason it only happens for _some_ SUBREV instructions is even more convoluted. It's because SIFoldOperands will sometimes shrink V_SUB_CO_U32_**e64** to V_SUBREV_CO_U32_**e32** even it does not manage to fold anything into it. This does seem wrong and is probably worth a closer look.

Reply:
OK, shrink pass has commuted it, but cannot shrink because of the sdst. It has marked the sdst with the allocation hint to be vcc and bailed until RA. Although it is possible to shink it because sdst is dead. Like this:
```
diff --git a/llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp b/llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp
index 4159dc694c1e..b20bcc987a7c 100644
--- a/llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp
+++ b/llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp
@@ -960,7 +960,7 @@ bool SIShrinkInstructions::runOnMachineFunction(MachineFunction &MF) {
       const MachineOperand *SDst = TII->getNamedOperand(MI,
                                                         AMDGPU::OpName::sdst);

-      if (SDst) {
+      if (SDst && !SDst->isDead()) {
         bool Next = false;

         if (SDst->getReg() != VCCReg) {
```
That patch has folded all 4 occurrences.

Reply:
https://github.com/llvm/llvm-project/pull/68028